### DISCUSSION TITLE
ESP32 with integrated OLED not waking up from deepsleep when coded with micropython

### DISCUSSION THREAD

**Discussion by bruno1950** [micropython/micropython#14044] [ESP32]
Hi, I'm trying to use an ESP32 with integrated OLED SSD1306 as a datalogger collecting data from BME280 and time from RTCDS3231 (since I'm planning to use it where wifi is often not working) and saving them on a SD card. I've realized on other ESP32s with m NO integrated OLED display the same project taht worked fine. Since this peculiar board when coded with Arduino IDE didn't work I flashed micropython and used this code that works but the board doesn't wake up just sending when it is supposed to awake the following message
ets Jul 29 2019 12:21:46

rst:0x5 (DEEPSLEEP_RESET),boot:0x17 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4728
load:0x40078000,len:14876
ho 0 tail 12 room 4
load:0x40080400,len:3368
entry 0x400805cc
MicroPython v1.21.0 on 2023-10-05; Generic ESP32 module with ESP32
Type "help()" for more information.
>>>   
Here is the code where for debugging purposes I chose a very short deepsleep (30 seconds)
'''
`from machine import deepsleep
from machine import Pin, SoftI2C
from ds3231_port import DS3231
import utime
import sys
import uos
import BME280
import ssd1306
from time import sleep

import machine, os, sdcard

# Assign chip select (CS) pin (and start it high)
cs = machine.Pin(15, machine.Pin.OUT)
# Intialize SPI peripheral (start with 1 MHz)
spi = machine.SPI(1,
                  baudrate=1000000,
                  polarity=0,
                  phase=0,
                  bits=8,
                  firstbit=machine.SPI.MSB,
                  sck=machine.Pin(25),
                  mosi=machine.Pin(14),
                  miso=machine.Pin(13))
# Initialize SD card
sd = sdcard.SDCard(spi, cs)
os.mount(sd, '/sd')

i2c = SoftI2C(scl=Pin(4), sda=Pin(5))
rtc = DS3231(i2c)
ds3231 = DS3231(i2c)
ds3231.get_time()
oled_width = 128
oled_height = 64
oled = ssd1306.SSD1306_I2C(oled_width, oled_height, i2c)

while True:
        oled.fill(0)
        bme = BME280.BME280(i2c=i2c)
        temperature = str(bme.temperature)
        humidity = str(bme.humidity)
        print('Temperature is ', temperature)
        print('Humidity is ', humidity)
        print('{:02d}:{:02d}:{:02d} '.format(utime.localtime()[3], utime.localtime()[4], utime.localtime()[5]), end='') 
        print('{:02d}:{:02d}:{:04d} '.format(utime.localtime()[2], utime.localtime()[1], utime.localtime()[0]), end='') 
        oled.text("Temp. "+temperature, 0, 0)
        oled.text("Hum, "+humidity, 0, 15)
        oled.show()
        sleep(5)
        oled.fill(0)
        file = open('/sd/test.txt', 'a')
        file.write(str(bme.temperature) + '; ' + str(bme.humidity) + '; '+ str(utime.localtime()[3:6]) + '; \n')
        file.close()
        oled.text('Awake', 0, 0)
        oled.text('but sleeping', 0, 10)
        oled.text('in 5 seconds', 0, 20)
        oled.text('for 30 seconds', 0, 30)
        oled.text(str(utime.localtime()[3:6]),0,45)
        oled.show()
        sleep(5)
        oled.fill(0)
        oled.poweroff()
        deepsleep(30000) 
'''
Thanks for your precious help: I'm an old(74 y) newbie and this maybe is a silly question.

**Comment by glenn20** [comment]
I'm not sure why your device is failing to wake from deepsleep, but I have a few comments that might help:

- after a deepsleep time completes, the device will do a full reboot, so, there is no need to wrap your updating code in `while True:`. 
- you often need to be careful about handling peripherals that will be powered down when entering deepsleep (see https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/system/sleep_modes.html).

I would try:

- test if deepsleep works on its own, without initialising the sd card or your external devices.
- add a short delay after oled.poweroff() and entering deepsleep,
- unmount the sd card before entering deepsleep
- check that none of your devices require any other deinitialisation steps before being entering deepsleep.

And finally, it is easier to help if you format code in your messages following the guidelines at: https://github.com/orgs/micropython/discussions/9111. You can edit your message to fix the formatting.

**Reply by bruno1950** [reply]
Hi, thx for your suggestions and apologies for not having formatted correctly the code ( I have used the <> of the upper bar but probably I made some mistakes)
I followed your suggestions and I used a very simple code with just a LED attached to the board ( I borrowed this code from an example reported in this site https://microcontrollerslab.com/micropython-esp32-deep-sleep-mode-wake-up-sources/  . 
Here you have the code
```
import machine
from machine import Pin
from time import sleep
led = Pin (2, Pin.OUT)   
led.value(1)                 
sleep(2)                   
led.value(0)               
sleep(2)                    
sleep(5)                
print('Setting to Deep Sleep Mode')
machine.deepsleep(10000)   
```

The code is correctly executed for the first run but again when the ESP32 is supposed to wake up this message appears and that's all
rst:0x5 (DEEPSLEEP_RESET),boot:0x17 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4728
load:0x40078000,len:14876
**ho 0 tail 12 room 4**   (Bold added to outline the difference)
load:0x40080400,len:3368
entry 0x400805cc
MicroPython v1.21.0 on 2023-10-05; Generic ESP32 module with ESP32
Type "help()" for more information.
>>> 

On the contrary in the example of the link there is a different printout re is the screenshot) 
![Screenshot 2024-03-08 140133](https://github.com/micropython/micropython/assets/121801771/2fe2f93a-aa69-49ff-bf94-c8158424c8e9)

However, I do not know if this difference can help to understand the problem and what it means.
Thanks again for your help

**Reply by davefes** [reply]
Which ESP32 device?

**Comment by bruno1950** [comment]
Hi, it's the one depicted in this link: https://randomnerdtutorials.com/esp32-built-in-oled-ssd1306/   However, I flashed Micropython also on a generic ESP32 (that I'm using without any problem with deepsleep and timer wake-up when coded with Arduino IDE)  and even this one shows the same behaviour. The code runs once
MPY: soft reboot
Setting to Deep Sleep Mode
then this message appears and nothing else happens.
ets Jul 29 2019 12:21:46

rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4728
load:0x40078000,len:14888
load:0x40080400,len:3368
entry 0x400805cc
MicroPython v1.22.2 on 2024-02-22; Generic ESP32 module with ESP32
Type "help()" for more information.

I know that I'm a newbie but this behaviour is absolutely awkward and I was not able to find any explanation. Waiting for your help. Thanks

**Reply by davefes** [reply]
It is a ESP32_WROOM variant.  On my ESP32-WROOM-32 putting your script into `main.py` gives:
```
rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4728
load:0x40078000,len:14888
load:0x40080400,len:3368
entry 0x400805cc
Setting to Deep Sleep Mode
ets Jun  8 2016 00:22:57

rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4728
load:0x40078000,len:14888
load:0x40080400,len:3368
entry 0x400805cc
Setting to Deep Sleep Mode
ets Jun  8 2016 00:22:57
```
Getting this:
```
MicroPython v1.22.2 on 2024-02-22; Generic ESP32 module with ESP32
Type "help()" for more information.
```
at the end suggests that "main" ran with no output.

How are you loading Micropython onto  the device?

**Comment by bruno1950** [comment]
Hi,

thanks for the prompt answer. I'm using Thonny and when I run a code without deep sleep it works. I've also tried to rename this code main.py but it just runs once, then prints the message, then nothing else

**Reply by davefes** [reply]
Unfortunately I don't use Thonny.  I use `esptool` for loading firmware and `rshell` to communicate with the board.  `mpremote` is the official command line tool.  

I will suggest installing esptool and run this command:
`esptool.py  --chip esp32 chip_id` (shows the chip reversion)
or hope someone who is familiar with Thonny appears on the scene.

Edit: if you search Thonny under ESP32 you will get lots of hits.

**Comment by bruno1950** [comment]
Hi, I tried to use esptool.py but I always get a SyntaxError and the source of the error should be esp32 ( since it's underlimed) However on the external envelop of my chip there is ESP32-WROOM- 32D. Could this help to identify the issue ? Thanks.

**Reply by davefes** [reply]
Did you install esptool on a Linux platform?  I use mostly ESP32-WROOM-32D and U variants.  I have only seen chip version 1 and version 3.  It may not help because the problem might be somewhere else.

I have read of people having problems running their device using Thonny and deepsleep, but have not really paid much attention as I don't use Thonny.  

I went back and looked at your pin assignments and then checked a file I use to initialise all the GPIO and I see some notes on it that might be relevant, ie pins 14 and 15.  For the ESP32 withut SPIRAM you can use pins 16 and 17.

The ports should be in these states after a re-boot, but to be really sure I do the following:

```py
# hardware init for ESP32
# set all GPIO to a known state after boot before any
# further imports

from umachine import Pin


pin0 = Pin(0, Pin.OUT, value=0)    # careful, strapping pin
pin2 = Pin(2, Pin.OUT, value=0)    # careful, strapping pin
pin4 = Pin(4, Pin.OUT, value=0)
pin5 = Pin(5, Pin.OUT, value=0)    # careful, strapping pin
pin12 = Pin(12, Pin.OUT, value=0)  # careful, strapping pin
pin13 = Pin(13, Pin.OUT, value=0)
pin14 = Pin(14, Pin.OUT, value=0)  # careful, goes hi during deepsleep
pin15 = Pin(15, Pin.OUT, value=0)  # careful, goes hi during deepsleep and strapping pin
pin18 = Pin(18, Pin.OUT, value=0)
pin19 = Pin(19, Pin.OUT, value=0)
pin21 = Pin(21, Pin.OUT, value=0)
pin22 = Pin(22, Pin.OUT, value=0)
pin23 = Pin(23, Pin.OUT, value=0)
pin25 = Pin(25, Pin.OUT, value=0)
pin26 = Pin(26, Pin.OUT, value=0)
pin27 = Pin(27, Pin.OUT, value=0)
pin32 = Pin(32, Pin.OUT, value=0)
pin33 = Pin(33, Pin.OUT, value=0)
pin34 = Pin(34, Pin.IN)            # input only, no pull-up or pull-down
pin35 = Pin(35, Pin.IN)            # input only, no pull-up or pull-down
```
I have seen suggestions not to use strapping pins but all my LoRa projects, based on a project on GitHub use several strapping pins.

After this script I will then setup the GPIO to what I really want.

I'll go and read the article on RANDOM Nerd

**Reply by davefes** [reply]
Have you got a link to a schematic for this board?

**Reply by davefes** [reply]
Read through the article and I see no mention of using deepsleep on that board.  Did you have that project working without deepsleep?

**Comment by bruno1950** [comment]
Hi, Answering your questions:
1) esptool installed on PC running with W11
2) this seems the pinout of my board 
![esp32-devkitC-v4-pinout](https://github.com/micropython/micropython/assets/121801771/d3697643-cfde-4a32-a5d1-ec7a5688c539)
3) yes, the codes work fine if I avoid deepsleep; for example if I just use sleep ( but it's the same as using delay in Arduino)

**Comment by bruno1950** [comment]
One more observation. I tried to modify the code adding a print statement and a second led that is supposed to blink if the code is executed after waking up and after on serial monitor appears the message "Waking up from deep sleep" and this happens but again just once.è That is, the first led blinks, then the board enters the deepsleep, then it seems to wake up since it prints the message  "Waking up from deep sleep" and the second led blinks but after that nothing else happens.
Here is the code modified
```
import machine
from machine import Pin
from time import sleep
k = 0
# Check if the ESP32 just woke up from deep sleep
if machine.reset_cause() == machine.DEEPSLEEP_RESET:
    print("Waking up from deep sleep")
    while k  < 5:
        led = Pin(15, Pin.OUT)  # GPIO14 as output for LED
        led.value(1)  # LED is ON
        sleep(1)  # Delay of 0.5 seconds
        led.value(0)  # LED is OFF
        sleep(1)
        k+=1
else:
    # Execute the LED blinking loop
    i = 0
    while i < 10:
        led = Pin(2, Pin.OUT)  # GPIO14 as output for LED
        led.value(1)  # LED is ON
        sleep(0.5)  # Delay of 0.5 seconds
        led.value(0)  # LED is OFF
        sleep(0.5)
        i += 1

    print("Setting to Deep Sleep Mode")
    sleep(5)
    machine.deepsleep(5000)  # 10000ms sleep time

```

**Reply by davefes** [reply]
Good morning,  The image I saw for your board was quite different https://randomnerdtutorials.com/esp32-built-in-oled-ssd1306/.  This one is the same as the boards I use.  First time that I have seen this image and I notice all the comments about GPIO state on every pin.

Have you got a ESP32 - DevKitC without a display on that you could run tests on?

**Reply by glenn20** [reply]
This code says to enter deepsleep ONLY if I did not wakeup from deepsleep. ie. after first boot, enter deepsleep, then after waking from deepsleep do NOT go back to sleep. If you unindent the last 3 lines it will repeatedly enter deepsleep.

**Reply by bruno1950** [reply]
Hi, I unindented the last three lines of the code as you suggested but still the code seems to run just once and when the deepsleep time has elapsed I got the usual message in the Thonny shell and nothing else happens. 
What I need is that the ESP32 when it wakes up run again the code, then enter again the deepsleep and so on. This is the modified code
```
import machine
from machine import Pin
from time import sleep

# Check if the ESP32 just woke up from deep sleep
if machine.reset_cause() == machine.DEEPSLEEP_RESET:
    print("Waking up from deep sleep")
else:
    # Execute the LED blinking loop
    i = 0
    while i < 10:
        led = Pin(18, Pin.OUT)  # GPIO14 as output for LED
        led.value(1)  # LED is ON
        sleep(0.5)  # Delay of 0.5 seconds
        led.value(0)  # LED is OFF
        sleep(0.5)
        i += 1
print("Setting to Deep Sleep Mode")
sleep(5)
machine.deepsleep(10000)  # 10000ms sleep time
```
and this is the message in the shell

>>> %Run -c $EDITOR_CONTENT
MPY: soft reboot
Setting to Deep Sleep Mode
ets Jul 29 2019 12:21:46

rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4728
load:0x40078000,len:14888
load:0x40080400,len:3368
entry 0x400805cc
MicroPython v1.22.2 on 2024-02-22; Generic ESP32 module with ESP32
Type "help()" for more information.
>>>

**Reply by robert-hh** [reply]
Thonny may not be a big help here. A simple terminal emulator or mpremote my suit better.

Since deepsleep causes a full reset of the device, you must ensure that your code is executed at boot time. So it must be part of main.py or imported by main.py. This means, it has to be permanently copied to the board's file system. When you execute Run in Thonny, the code is just copied to RAM and gone after deepsleep wakeup.

**Comment by bruno1950** [comment]
Hi,

good evening ( here in Turin, Italy, it's 9.10 pm )

Yes I flashed micropython also on a generic ESP32 that I've used several times with no problems for datalogging and deep sleep.

I got, as I told you, the same result that I had with trhe Wemos Lolin 32 with integrated OELD that was the first board on which I tried micropython.

However, I found an interesting phenomenon: if, instead of using deep sleep I use machine.lightsleep, the board wakes up and runs the code without any problem.

So I am forced to think that the problem is in the deepsleep function when applied to ESP32. And this fact sounds very strange for me because I found examples on the web where deep sleep is implemented into an ESP32 without problems.

One final word: I really appreciate your patience and your help: as an old(74) newbie I must confess that I've found the forums not so kind with old men( especially when they are newbies) so that I got the wrong  idea that Arduino and Python forums were No Country for Old Men, as the movie says...

**Reply by davefes** [reply]
Just to clarify, as I am 5 years older than you, the last script you posted fails when you run it on a generic ESP32 WITHOUT an attached display.  Correct?

**Comment by bruno1950** [comment]
Hi,

now I understand why you are so patient...  

However, yes you are correct : with a generic ESP32 ( with esp32-WROOM-32D chip) I got the same problem that I had with the Wemos Lolin 32 with integrated OLED.

However, the code runs without problem if I use lightsleep anot deepsleep ( obviously it works if I use just sleep).

**Reply by davefes** [reply]
Print out the reset_cause each time.  When I tested one of your earlier scripts I would sometimes get a different reset_cause.  Also, that pin15, maybe try another pin.

For lightsleep you need to hold GPIO state before going to sleep and then "unhold" after coming out of lightsleep.

Edit: reading the latest documents I see you can also do:
```py
# enable pad hold in deep-sleep for non-RTC GPIO
esp32.gpio_deep_sleep_hold(True)
```

**Reply by davefes** [reply]
Just went back to retry your earlier script as I thought I saw some RTC type resets rather than DEEPSLEEP resets and can't reproduce them.

Then I thought ... my test board has a few added components so it may not be a fair test.  I always put a 1uF on the enable line and this board has 1000uF on 3V3 to help cope with WiFi current surges ... well that is the theory.

Later today I can make sure that my board is setup exactly the same as yours.

**Comment by bruno1950** [comment]
Thanks for your efforts. Also my board as a 10 uF between EN pin and GND just to avoid pressing BOOT button to upload codes : no other things attached except a LED with a 220 ohm resistor

I've changed the ledpin according to your suggestion and this is the modified code
```
import machine
from machine import Pin
from time import sleep
from machine import deepsleep

# Check if the ESP32 just woke up from deep sleep
if machine.reset_cause() == machine.DEEPSLEEP_RESET:
    print("Waking up from deep sleep")
else:
    # Execute the LED blinking loop
    i = 0
    while i < 10:
        led = Pin(18, Pin.OUT) 
        led.value(1)  # LED is ON
        sleep(0.5)  # Delay of 0.5 seconds
        led.value(0)  # LED is OFF
        sleep(0.5)
        i += 1

    print("Setting to Deep Sleep Mode")
    machine.deepsleep(10000)  # 10000ms sleep time

and this the message at the reset
ets Jul 29 2019 12:21:46

rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4728
load:0x40078000,len:14888
load:0x40080400,len:3368
entry 0x400805cc
MicroPython v1.22.2 on 2024-02-22; Generic ESP32 module with ESP32
Type "help()" for more information.
>>> 
```Have a nice day

**Reply by davefes** [reply]
Does "Have a nice day" mean that it is working now or do still only get one loop?

**Reply by davefes** [reply]
Just checking ... are there 3 pin devices mounted to the right side of the 3V3 regulator and the USB/serial chip?  Try reducing the 10uF to 1uF.

**Comment by bruno1950** [comment]
No, 
I've just said "have a nice day" because from your previous post I thought that you were engaged in other activities. Still only one loop. 
Now I'm going to sleep (hopefully better than esp32) and tomorrow I'll try again with your suggestion to change the capacitor ( I have to find one) 
--
Inviato da Libero Mail

**Reply by davefes** [reply]
I am "multi-tasking", ie domestic activities and Micropython.

**Comment by shariltumin** [comment]
This is the logical structure of your code:
```py
if machine.reset_cause() == machine.DEEPSLEEP_RESET:
    print("Waking up from deep sleep")
else:
    ...
    print("Setting to Deep Sleep Mode")
    machine.deepsleep(10000)  # 10000ms sleep time
```

which means it will go into deepsleep once. After this deepsleep it will ```print("Waking up ....")``` and exit from main.py to the REPL. If you do a soft reset ctrl-d in the REPL, it will enter the else block and do the deepsleep again.

This is probably not what you want. You want to go into deep sleep every time your code runs.

Try restructuring your code:

```py
...
if machine.reset_cause() == machine.DEEPSLEEP_RESET:
    print("Waking up from deep sleep")

# do some work here
print('do some work')

# go to deepsleep
print('Countdown to new deepsleep unless ctrl-c')
print('You have 10 seconds before deep sleep')
c=10
while c>=0:
  try:
     print(f'{c}..', end='')
     sleep(1)
     c-=1
  except:
     print(f'\nabort countdown at {c}')
     print('Back to REPL - this code run from main.py')
     break
else:
  print('\ngoing to deepsleep now!!')
  mc.deepsleep(1000)
  print('You will never see this line printed out')
```

This code runs from main.py, the countdown makes it possible to break out of main.py into the REPL. This is important if you want to edit or delete main.py.

**Reply by davefes** [reply]
Good point ... I didn't try running any of the later scripts as his test code from 2 days ago wouldn't run properly on his device.

```py
import machine
from machine import Pin
from time import sleep

led = Pin (2, Pin.OUT)   
led.value(1)                 
sleep(2)                   
led.value(0)               
sleep(2)                    
sleep(5)                
print('Setting to Deep Sleep Mode')
machine.deepsleep(10000)   
```

**Comment by bruno1950** [comment]
Hi, at the end I succeded in obtaining what I need: the ESP32 wakes up, runs the code and goes to deepsleep, then wakes up, runs again the code and so on. I've tried with an ESP32 without integrated OLED but I connected a 1306SSD OLED through i2c and it works. 
I unindented the last three lines of the code as suggested by Glenn 20 ( I don't understand why the first attempt was unsuccesful) 
Thanks to you all for your help. Have a nice Sunday.

**Comment by bruno1950** [comment]
Hi, thanks for your suggestion. Indeed now I renamed the code main.py. As a newbie I started with Thonny since I come from Arduino IDE. I'll try to use what you're suggesting . Have a nice day 
--
Inviato da Libero Mail
